server/src/main/java/com/google/fhir/gateway/AuditEventHelperImpl.java

```diff
@@ -0,0 +1,365 @@
+  }
+
+  // TODO handle the case for Bundle operations
+  @Override
+  public void processAuditEvents() {
+    try {
+      List<AuditEvent> auditEventList = new ArrayList<>();
+
+      List<IBaseResource> resources = extractFhirResources(responseContent);
+
+      switch (this.requestDetailsReader.getRestOperationType()) {
+        case SEARCH_TYPE:
+        case SEARCH_SYSTEM:
+        case GET_PAGE:
+          auditEventList =
+              createAuditEventCore(
+                  resources, BalpProfileEnum.PATIENT_QUERY, BalpProfileEnum.BASIC_QUERY);
+          break;
+
+        case READ:
+        case VREAD:
+          auditEventList =
+              createAuditEventCore(
+                  resources, BalpProfileEnum.PATIENT_READ, BalpProfileEnum.BASIC_READ);
+          break;
+
+        case CREATE:
+          auditEventList =
+              createAuditEventCore(
+                  resources, BalpProfileEnum.PATIENT_CREATE, BalpProfileEnum.BASIC_CREATE);
+          break;
+
+        case UPDATE:
+          auditEventList =
+              createAuditEventCore(
+                  resources, BalpProfileEnum.PATIENT_UPDATE, BalpProfileEnum.PATIENT_UPDATE);
+          break;
+
+        case DELETE:
+
+          // NOTE: The success of processing of DELETE is heavily dependent on server validation
+          // policy e.g. If you have a permission list that references the resource being deleted,
+          // it breaks the delete operation. Likewise if you reference the resource in AuditEvent
+          // such as logging when you first created or updated it the operation breaks. In case of
+          // such an error, AuditEvent.outcome and AuditEvent.outcomeDesc are populated accordingly
+
+          // Also note, this implementation only fully captures that a specific resource was deleted
+          // and by whom but not the compartment owner.
+          // With no access to the actual deleted resource we can't get the compartment owner unless
+          // the resource itself is a Patient. A crude way to get the owner would be to fetch the
+          // actual resource from the database first before creating the AuditEvent.
+
+          // This implementation skips logging Deletes if the operation is conditional (thus no
+          // Resource ID present)
+
+          auditEventList =
+              createAuditEventCore(
+                  resources, BalpProfileEnum.PATIENT_DELETE, BalpProfileEnum.BASIC_DELETE);
+          break;
+
+        default:
+          break;
+      }
+
+      // TODO Investigate bulk saving (batch processing) instead to improve performance. We'll
+      // probably need a mechanism for chunking e.g. 100, 200, or 500 batch
+      for (AuditEvent auditEvent : auditEventList) {
+        auditEvent.getPeriod().setEnd(new Date());
+        this.httpFhirClient.postResource(auditEvent);
+      }
+    } catch (IOException exception) {
+      logger.error(exception.getMessage(), exception);
```

Comment:
What is the rationale for not throwing an exception in this case? Is it because we want the server to still succeed when posting AuditEvent resources fail?

Reply:
Yeah that was the idea.

I've refactored this a bit. The exception caught here is the IOException from the `httpFhirClient` that can be thrown after a `postResource` operation. 

I've narrowed down the scope to per operation and thrown + logged the exception using the gateway's `ExceptionUtil` class. Now all resources up to when an exception is throw will have been processed successfully.